R/anomaly_detection.R

```diff
@@ -0,0 +1,106 @@
+
+#' NSE version of do_anomaly_detection_
+#' @export
+do_anomaly_detection <- function(df, time, value, ...){
+  time_col <- col_name(substitute(time))
+  value_col <- col_name(substitute(value))
+  do_anomaly_detection_(df, time_col, value_col, ...)
+}
+
+#' Detect anomaly data
+#' @param df Data frame
+#' @param time_col Column that has time data
+#' @param value_col Column that has value data
+#' @param direction Direction of anomaly. Positive ("posi"), Negative ("neg") or "both".
+#' @param e_value Whther expected values should be returned.
+#' @param ... extra values to be passed to AnomalyDetection::AnomalyDetectionTs.
+#' @export
+do_anomaly_detection_ <- function(df, time_col, value_col, direction="both", e_value=TRUE, ...){
+  loadNamespace("dplyr")
+  loadNamespace("AnomalyDetection")
+
+  if(!direction %in% c("both", "pos", "neg")){
+    stop("derection must be 'both', 'pos' or 'neg'")
+  }
+
+  pos_anom_col <- avoid_conflict(colnames(df), "pos_anomaly")
+  pos_val_col <- avoid_conflict(colnames(df), "pos_value")
+  neg_anom_col <- avoid_conflict(colnames(df), "neg_anomaly")
+  neg_val_col <- avoid_conflict(colnames(df), "neg_value")
+  exp_val_col <- avoid_conflict(colnames(df), "expected_value")
+
+  grouped_col <- grouped_by(df)
+
+  do_anomaly_detection_each <- function(df){
+    if(!is.null(grouped_col)){
+      # drop grouping columns
+      df <- df[, !colnames(df) %in% grouped_col]
+    }
+
+    # AnomalyDetection::AnomalyDetectionTs expects
+    # a data frame whose first column is time column
+    # and second column is value column, so it's created
+    data <- df[, c(time_col, value_col)]
+    # time column should be posixct, otherwise AnomalyDetection::AnomalyDetectionTs throws an error
+    data[[time_col]] <- as.POSIXct(data[[time_col]])
+
+    # this will be overwritten by expected values
+    expected_values <- df[[value_col]]
+
+    if(direction == "both" || direction == "pos"){
```

Comment:
positive case and negative case seems to be almost same code.
Would it be possible to share same code between them?